designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,963 @@
+
+This information allows users to make logical decisions about how the config should be constructed.
+
+A configuration function may return an object or an array of objects. An error is thrown if any other type of value is returned.
+
+#### Checking for Rule Existence
+
+One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:
+
+```js
+const requireIndex = require("requireindex");
+
+module.exports = (context) => {
+    const myConfig = {
+        files: ["*.js"],
+        plugins: {
+            custom: {
+                rules: requireIndex("./custom-rules")
+            }
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+};
+```
+
+#### Including Function Configs in an Array
+
+A function config can be used anywhere a config object or a config array is valid. That means you can insert a function config as a config array member:
+
+```js
+module.exports = [
+    (context) => someObject,
+    require("eslint-config-myconfig")
+];
+```
+
+Each function config in an array will be executed with a `context` object when ESLint evaluates the configuration file. This also means that shareable configs can export a function instead of an object or array.
+
+**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.
+
+### The `@eslint/eslintrc` Utility
+
+To allow for backwards compatibility with existing configs and plugins, an `@eslint/eslintrc` utility is provided. The package exports the following classes:
+
+* `ESLintRCCompat` - a class to help convert `.eslintrc`-style configs into the correct format.
+
+
+```js
+class ESLintRCCompat {
+    
+    constructor(baseDir) {}
+    
+    config(configObjct) {}
+    plugins(...pluginNames) {}
+    extends(...sharedConfigNames) {}
+    env(envObject) {}
+
+}
+```
+
+#### Importing existing configs
+
+The `ESLintRCCompat#extends()` function allows users to specify an existing `.eslintrc` config location in the same format that used in the `.eslintrc` `extends` key. Users can pass in a filename, a shareable config name, or a plugin config name and have it converted automatically into the correct format. For example:
+
+```js
+const { ESLintRCCompat } = require("@eslint/eslintrc");
+
+const eslintrc = new ESLintRCCompat(__dirname);
+
+module.exports = [
+    "eslint:recommended",
```

Comment:
Does that translate internally to something like the following

```js
module.exports = [
  {
    plugins: {
      'eslint:recommended': require('eslint/path/to/internal/rules')
    }
  }
]
```

Reply:
No, plugin declarations create namespaces, not definitions. The string value would just need to be replaced with an object during normalization.

Reply:
That's what I'm talking about. To what will be normalized the `'eslint:recommended'` string?